src/custom-lists/background/storage.ts

```diff
@@ -0,0 +1,191 @@
+import { ListObject, PageObject } from './types'
+
+const COLLECTION_NAME = 'customLists'
+const PAGE_LIST_ENTRY = 'pageListEntries'
+
+// TODO: Add typings for the class
+export default class CustomListStorage extends FeatureStorage {
+    constructor({ storageManager }) {
+        super(storageManager)
+
+        this.storageManager.registerCollection(COLLECTION_NAME, {
+            // different version for adding a new table.
+            version: new Date(2018, 6, 12),
+            fields: {
+                id: { type: 'string', pk: true },
+                name: { type: 'text' },
+                isDeletable: { type: 'binary' },
+                isNestable: { type: 'binary' },
+                createdAt: { type: 'datetime' },
+            },
+            indices: [
+                { field: 'id', pk: true, autoInc: true },
+                { field: 'name', unique: true },
+                { field: 'isDeletable' },
+                { field: 'isNestable' },
+                { field: 'createdAt' },
+            ],
+        })
+
+        this.storageManager.registerCollection(PAGE_LIST_ENTRY, {
+            // different version for adding a new table.
+            version: new Date(2018, 6, 12),
+            fields: {
+                listId: { type: 'string' },
+                pageUrl: { type: 'string' },
+                fullUrl: { type: 'string' },
+                createdAt: { type: 'datetime' },
+            },
+            indices: [
+                { field: ['listId', 'pageUrl'], pk: true },
+                { field: 'listId' },
+                { field: 'pageUrl' },
+            ],
+        })
+    }
+
+    // Return all the list in the DB
+    //  TODO: Use pagination if required and decide correct types
+    async fetchAllList({ query = {}, opts = {} }: { query?: any; opts?: any }) {
+        const x = await this.storageManager.findAll(COLLECTION_NAME, query, opts)
+        return this.changeListsbeforeSending(x, [])
+    }
+
+    // Takes the list as they come from Db and does some pre-processing before sending.
+    async changeListsbeforeSending(lists: object[], pageEnteries: object[]) {
+        const mappedLists = lists.map((list: ListObject) => {
+            const page = pageEnteries.find(({ listId }: PageObject) => listId === list.id)
+            delete list["_name_terms"]
+            return {
+                ...list,
+                // Set an empty pages array for manupulating in redux store.
+                pages: [],
+                active: Boolean(page),
+            }
+        })
+        return mappedLists
+    }
+
+    async fetchListById(id: Number) {
+        const list = await this.storageManager.findObject(COLLECTION_NAME, { id })
+        const pages = await this.storageManager.findAll(PAGE_LIST_ENTRY, { listId: list.id })
+        delete list["_name_terms"]
+        return {
+            ...list,
+            pages: pages.map((page: PageObject) => page.fullUrl),
```

Comment:
All the StorageManager methods should have a generic type param. So what that means is, to avoid putting these typings all over the place, you should be able to do something like this on your method calls:

```ts
const list = await this.storageManager.findObject<ListObject>(COLLECTION_NAME, { id })
const pages = await this.storageManager.findAll<PageObject>(PAGE_LIST_ENTRY, { listId: list.id })
```

Then your `list` and `pages` variables should be automatically typed properly as `ListObject` and `PageObject[]`, respectively (and all subsequent ops, like array mapping, should also get those types automatically). Let me know if it doesn't work though

Reply:
works :+1: